giFT-foo/OpenFT.conf.5
..
.de TQ
.br
.ns
.TP \\$1
..
.TH OpenFT.conf 5 "13 April 2003" "OpenFT version 0.0.8-2" 
.SH NAME
OpenFT.conf \- configuration file for OpenFT, used by
.BR giFT (1)
.SH DESCRIPTION
.B OpenFT.conf
uses a very basic configuration file format. It's divided in sections, starting
with
.RB [ sectionname "]."
parameters are set using 
.RB ` parameter " = " value '.
Lines starting with 
.B #
are comments.
.P
The file must be located at
.IR ~/.giFT/OpenFT/OpenFT.conf .
.SH OPTIONS
.SS [main]
Main options
.TP
.B class
Node classification.  This value is very important and will determine your
role on the network.
.IP
The following settings are allowed:
.IP
.IR 1 " USER"
.br
.IR 3 " SEARCH"
.br
.IR 5 " INDEX"
.br
.IR 7 " SEARCH & INDEX"
.IP
If you unsure it is highly recommended that you select 1.  More information
may be found by consulting other documentation.
.IP
.I "Default: "
1
.TP
.B port 
Core listening port.  OpenFT uses a compacted binary protocol for the core
communication with other nodes.  If this machine is unable to provide both
port and http_port to the outside directly (blocked by either by a firewall
or a NAT setup) you should select 0 here.
.IP
.BI NOTE:
If you are unable to route this port (along with the HTTP port) you will be
unable to download from other users with that same inability.  You will,
however, maintain the ability to upload to other non-firewalled users.
Selecting non-zero when the port is not available to the outside will cause
very broken behaviour, namely search results will be extremely unreliable.
.IP
.I "Default: "
1215
.TP
.B http_port
HTTP server port.  OpenFT's implementation also makes use of HTTP for all
file transfers.  See above for more details.  Please provide a valid port
here regardless of your firewall state.
.IP
.I "Default: "
1216
.TP
.B alias
Arbitrary name you'd like to associate with your node.  This field is
optional and it is NOT to be trusted.  The name is not unique to the
network.  Please do not choose useless aliases.
.IP
.I "Default: "
none
.SS [nodepage]
Nodepage options
.TP
.B private
If set non-zero, OpenFT will not allow remote hosts to access the nodepage.
This option also depends on local/hosts_allow shown below. For the
interest of debugging I suggest you just leave this as is.
.IP
.I "Default: "
0
.SS [search]
Search node settings
.TP
.B children
Maximum number of "child" nodes you are able to support.  You should not
touch this unless you know what you're doing.
.IP
.BI NOTE:
This option has no effect unless you configure as a SEARCH node.
.IP
.I "Default: "
500
.TP
.B env_path
Absolute path where the search database environment should be stored.  Make
sure to provide a path with plenty of hard disk space (~100-300MB) and on a
device with efficient hard drive access to improve swap time.
.IP
.I "Default: "
~/.giFT/OpenFT/db
.TP
.B env_cache
Total cache size for the entire search database environment.  The idea is
to find a perfect balance here between disk access and memory usage, which
we are still widely experimenting with.  If you have a lot of RAM to
dedicate to your search node, by all means increase this number accordingly
as it will improve overall performance.
.IP
.BI WARNING:
Specifications less than 500MB (and we expect you will specify less than
that) are subject to a 20% inflation due to some libdb implementation
detail.
.IP
.BI NOTE:
Units are specified in bytes, and as such you will not be allowed to
specify 1GB or higher.  While libdb supports this, the API is more obscure
than this simple configuration option, hopefuilly the configuration will
be extended at some point to support this.
.IP
.I "Default: "
83886080
.TP
.B env_priv
If the database environment is private, you will be unable to use the
database tools provided by the libdb util package.  This will make it
almost to gain memory pool statistics for the entire process and makes
tweaking of the search code almost impossible.  However, most users will
want this turned on as it will create a rather large temporary swap file
in the database directory otherwise.
.IP
.I "Default: "
0
.SS [local]
LAN mode facilities
.TP
.B lan_mode
OpenFT offers a special configuration for users who wish to operate a
network from within a local area network (LAN).  This mode causes giFT to
reject authorization from any non-local address as well as advertising it's
presence via the local UDP broadcast.  If this switch is non-zero you will
not be able to contact the main network.
.IP
.I "Default: "
0
.TP 
.B hosts_allow
When LAN mode is enabled this configuration switch is used to
verify which hosts are allowed connection to the core OpenFT port.  It is
highly recommended that you just don't touch this, regardless of what
lan_mode's value is.
.IP
.BI NOTE:
This switch provides a very crude loophole that could allow you to create
a "private" OpenFT network between a list of known nodes.  Using this
"feature" is going to be quite cumbersome as it is intended to be used
by developers only.  You have been warned.
.IP
.I "Default: "
LOCAL
.SH AUTHOR
Josh Guilfoyle <jasta@gotlinux.org>. Made into a manpage by Eelco Lempsink
<eelco@33lc0.net>
.SH "SEE ALSO"
.BR giFT (1), 
.BR gift.conf (5),
http://gift.sourceforge.net/docs.php


giFTcurs/NEWS
New in 0.6.0, not released yet:
	* Now uses GLib 2.
	* The md5file option is now deprecated, since giFT verifies the
	  checksum of downloads itself.
	* Banned words are stored in the config file.

New in 0.5.7, released 2003-05-01:
	* Ability to exclude search terms by placing a '-' before them.
	* Sorting of hits and transfers by arbitrary fields which can be
	  specified in the config file.

New in 0.5.6, released 2003-02-27:
	* Improved format strings.
	* 'q' exits the program.
	* Portability fixes.
	* Don't crash when a search result without a mime field is
	  to be shown. Thanks to Jan-Marek Glogowski for discovering
	  and fixing this.

New in 0.5.5, released 2002-11-25:
	* Captital E in the transfer screen (un)expands all transfers.
	* Slight protocol change caused transfers to not be seen.
	* Minor fixes.

New in 0.5.4, released 2002-10-26:
	* Formats for the search results were broken.
	* Small fix for transfers when giFT dies.

New in 0.5.3, released 2002-10-25:
	* Layout is determined by configurable formatting strings.
	* Better support for viewing important messages from giFT.
	* Support for Tru64 (OSF1 V4.0) and OSF1 curses.
	* Various bug fixes.

New in 0.5.2, released 2002-09-23:
	* Smashed a few bugs.
	* Meta data shows in the Extra Information box. Use j/k to scroll it.

New in 0.5.1, released 2002-09-12:
	* Faster hit sorting algorithm.
	* Small interface protocol change.
	* Transfers are sorted, largest to smallest.
	* Automatic awakening moved to a separate binary.

New in 0.5.0, released 2002-08-25:
	* Resizing finally works right away. This was the oldest bug.
	* Interface protocol changed drastically.

New in 0.4.3, released 2002-07-07:
	* A button bar for the very laziest of users. Also good for new users,
	  so the default message timer has been removed. May be disabled, see
	  giFTcurs.conf(5).
	* More things are clickable, check the ChangeLog.
	* The 'X' key was changed to 'T', to make cancelling transfers harder.

New in 0.4.2, released 2002-06-26:
	* Selections now have their own color. The old behaviour of inverting
	  the color is still available, see giFTcurs.conf(5).
	* Two new sorting orders.
	* A setting that will ignore hits with zero availability.
	* A compile fix for FreeBSD and one for Mac OS X.
	* Support for the new pause feature in giFT. (Press 'p' on a download.)

New in 0.4.1, released 2002-05-20:
	* An option to log hash and filename of started downloads to a file,
	  so that a user may run md5sum -c.
	* A PowerPC-specific segfault was fixed.

New in 0.4.0, released 2002-05-18:
	* Search results and downloads are expandable trees.
	* A new transfer style. The original is still available, see
	  giFTcurs.conf(5) for more information.
	* A new manpage, giFTcurs.conf(5).
	* Support for chunk reporting.
	* Bug fixes.

New in 0.3.4, released 2002-04-23:
	* libgpm support
	* Search history.
	* Very much improved settings code.
	* Temporary workaround for a protocol change.

New in 0.3.3, released 2002-04-18:
	* Improved settings screen.
	* Proper mouse support.
	* / for searching among results.
	* Show average bandwidth on completed transfers.
	* Brazilian Portuguese, Danish and Dutch translations.
	* Files in the download list are marked in the search results list.
	* Mouse wheel support in terminals which support it.
	* Ban words filter on incoming search results.

New in 0.3.2, released 2002-03-20:
	* Comply with a procotol change that affected downloading.
	* Alphabetical sorting of results.
	* Nodepage uploads are automagically removed when complete.
	* A few bugs were fixed.

New in 0.3.1, released 2002-03-13:
	* Stats work again.
	* A screen for changing colors was added.
	* The transfer screen was improved.
	* A partial French translation was added.
	* Lots of bugs were fixed.

New in 0.3.0, released 2002-02-27:
	* Adhere to the new interface protocol.
	* Most likely stuff we've forgotten about.

New in 0.2.1, released 2002-01-17:
	* Many small improvements and bug fixes.

New in 0.2.0, released 2002-12-19:
	* Clean hit list automagically on new search.
	* Timed delays between hit list update, no CPU hog here.
	* Downloads through giFT.
	* The new somewhat limited <stats/> is used and the main
	  screen was adapted to this.
	* Fixed a socket leak and some memory leaks.

New in 0.1.0, released 2001-11-20:
	* Initial release.


giFT.oldish/OpenFT/openft.c
/*
 * openft.c
 *
 * Copyright (C) 2001-2002 giFT project (gift.sourceforge.net)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 */

/* hack hack hack */
#define __OPENFT_C__

#include <signal.h>

#include "openft.h"

#include "netorg.h"

#include "share_comp.h"
#include "search.h"
#include "daemon.h"

#include "xfer.h"
#include "html.h"

#include "file.h"
#include "parse.h"

#include "md5.h"

/*****************************************************************************/

#define MAINTAIN_LINKS_TIMER (2 * MINUTES)

/*****************************************************************************/

int openft_shutdown           = FALSE; /* boolean that is set when OpenFT is
                                        * shutting itself down */

/*****************************************************************************/

Config *openft_conf           = NULL;  /* ~/.giFT/gift.conf */

Connection *ft_self           = NULL;  /* current class information */
Connection *http_bind         = NULL;

Protocol *openft_proto        = NULL;  /* OpenFT protocol plugin pointer */

static unsigned long ft_maint = 0;     /* timer for connection maintenance */

/*****************************************************************************/

/* handle share actions for giFT */
static void ft_daemon_share (FileShare *file, ProtocolCommand command,
                             void *data)
{
	switch (command)
	{
	 case PROTOCOL_SHARE_ADD:    ft_share_local_add (file);    break;
	 case PROTOCOL_SHARE_REMOVE: ft_share_local_remove (file); break;
	 case PROTOCOL_SHARE_FLUSH:  ft_share_local_flush ();      break;
	 case PROTOCOL_SHARE_SYNC:   ft_share_local_sync ();       break;
	 default:                                                  break;
	}
}

/*****************************************************************************/

static Connection *cleanup_conn (Connection *c, Node *node, void *data)
{
	node_disconnect (c);

#if 0
	openft_share_remove_by_host (node->ip, TRUE);
#endif

	return NULL;
}

/* shutdown */
static void ft_daemon_destroy (Protocol *p)
{
	TRACE_FUNC ();

	hash_algo_unregister (p, "MD5");

	/* boolean used by node_disconnect to disable certain behavior */
	openft_shutdown = TRUE;

	/* write nodes cache */
	node_update_cache ();

	/* disconnect */
	conn_clear ((ConnForeachFunc) cleanup_conn);

	/* cleanup */
	config_free (openft_conf);

	timer_remove (ft_maint);

	net_close (http_bind->fd);
	net_close (ft_self->fd);

	node_free (http_bind);
	node_free (ft_self);

	/* get rid of our protocol specific data */
	ft_share_local_cleanup ();
}

/*****************************************************************************/

static Connection *http_start (unsigned short port)
{
	Connection *c;

	TRACE (("http_port = %hu", port));

	c = connection_new (openft_proto);
	c->fd = net_bind (port, FALSE);

	input_add (openft_proto, c, INPUT_READ,
	           (InputCallback) http_server_incoming, FALSE);

	node_conn_set (ft_self, 0, -1, port, NULL);

	return c;
}

static Connection *ft_start (unsigned short klass, unsigned short port)
{
	Connection *c;

	/* firewalled users can not go beyond the user class */
	if (port == 0)
		klass = NODE_USER;

#ifndef USE_LIBDB
	if (klass & NODE_SEARCH)
	{
		GIFT_WARN (("dropping NODE_SEARCH status due to lacking libdb "
		            "support"));
	}

	klass &= ~NODE_SEARCH;
#endif /* !USE_LIBDB */

	TRACE (("class = %hu, port = %hu", klass, port));

	c = node_new (net_bind (port, FALSE));

	input_add (openft_proto, c, INPUT_READ,
	           (InputCallback) ft_handle_incoming, FALSE);

	node_state_set (c, NODE_CONNECTED);
	node_class_set (c, klass);
	node_conn_set  (c, 0, port, -1, NODE_ALIAS);

	if (!(NODE (c)->cap))
		NODE (c)->cap = dataset_new (DATASET_LIST);

	return c;
}

static unsigned char *openft_hash_md5 (char *path, char *type, int *len)
{
	if (len)
		*len = 16;

	return md5_digest (path, 0);
}

/*****************************************************************************/

/*
 * OpenFT was parsed out of /path/to/libOpenFT.so
 */
int OpenFT_init (Protocol *p)
{
	int klass;
	int port;
	int http_port;

	TRACE_FUNC ();

	hash_algo_register (p, "MD5", (HashAlgorithm) openft_hash_md5);

	openft_proto = p;

	/* set support */
	dataset_insert (&p->support, "range-get",  10, I_PTR (TRUE));
	dataset_insert (&p->support, "chat-user",  10, I_PTR (FALSE));
	dataset_insert (&p->support, "chat-group", 11, I_PTR (FALSE));

	/* setup the symbols for the daemon to call */
	p->callback   =                    ft_daemon_callback;
	p->destroy    =                    ft_daemon_destroy;
	p->download   = (ProtocolCallback) ft_daemon_download;   /* xfer.c */
	p->upload     = (ProtocolCallback) ft_daemon_upload;     /* xfer.c */
	p->source_cmp =                    ft_daemon_source_cmp; /* xfer.c */
	p->share      = (ProtocolCallback) ft_daemon_share;

	/* setup the configuration table */
	openft_conf = gift_config_new ("OpenFT");
	port        = config_get_int (openft_conf, "main/port=1215");
	http_port   = config_get_int (openft_conf, "main/http_port=1216");
	klass       = config_get_int (openft_conf, "main/class=1");

	/* startup the listening sockets */
	ft_self   = ft_start   (klass, port);
	http_bind = http_start (http_port);

	/* add the capabilities */
#ifdef USE_ZLIB
	dataset_insert (&NODE (ft_self)->cap, "ZLIB",     5, STRDUP ("ZLIB"));
#endif /* USE_ZLIB */
	dataset_insert (&NODE (ft_self)->cap, "MD5-FULL", 9, STRDUP ("MD5-FULL"));

	/* setup the connection maintenance timer (2 minute intervals) */
	ft_maint = timer_add (MAINTAIN_LINKS_TIMER,
	                      (TimerCallback) node_maintain_links, NULL);

	/* make sure ~/.giFT/OpenFT/shares.gz is up to date */
	share_comp_write ();

	/* delete any host shares left over from previous search node sessions */
	file_rmdir (gift_conf_path ("OpenFT/db"));

	/* OpenFT, here we come :) */
	node_maintain_links (NULL);

	return TRUE;
}
